File: lab3v17/vehicle.py
import abc

class Vehicle(abc.ABC):
    def __init__(self, fuel, distance):
        self.fuel = fuel
        self.__distance = distance

    @property
    def get_distance(self):
        return self.__distance
    
    @get_distance.setter
    def set_distance(self, distance):
        self.__distance = distance

    @abc.abstractmethod
    def fuel_consumption_calc(self):  # абстрактний
        pass

    # Віртуальний метод
    def info(self):
        return f"Транспорт з загальним запасом палива {self.fuel} Л"

    def __del__(self):
        print(f'Destructor called, {self.__class__.__name__} deleted.')

    def __str__(self):
        return f"Витрати транспорту на {self.get_distance} km: {self.fuel_consumption_calc()} liters"


class Bus(Vehicle):
    def __init__(self, fuel, distance): 
        super().__init__(fuel, distance)

    def fuel_consumption_calc(self):  # override (абстрактного)
        return self.fuel * self.get_distance / 100

    # Override (віртуального) 
    def info(self):
        return "Автобус для пасажирів"


class Truck(Vehicle):
    def __init__(self, fuel, distance):
        super().__init__(fuel, distance)

    def fuel_consumption_calc(self):  # override (абстрактного)
        return self.fuel * self.get_distance / 100

File: lab3v17/vehicle.cpp
#include <iostream>

class Vehicle //загальний класс
{
public:
    Vehicle( double fuel, double distance) //загальний конструктор з параметрами
    {
        m_fuel = fuel;
        m_distance = distance;
    }
    double  GetFuel(){return m_fuel;}
    double  Getdistance(){return m_distance;}

    void SetFuel(double fuel){m_fuel = fuel;}
    void Setdistance(double distance){m_distance = distance;}

    virtual double fuel_consumption_calc() = 0; //абстрактний метод(без реалізації)

    virtual void info() // віртуальний метод(з реалізацією, яку можна змінити)
    {
        std::cout <<"total fuel content" << m_fuel << std::endl; 
    }

    ~ Vehicle() // деструктор
    {
        std::cout <<"Destructor called" << std::endl;
    }
    
protected:
    double m_fuel;
    double m_distance;
};

class Bus: public Vehicle // класс наслідник №1
{
public:
    Bus(double fuel, double distance) : Vehicle( fuel, distance ) {} // заміна base()

    double fuel_consumption_calc() override //override абстрактного методу
    {
       return m_fuel * m_distance / 100;
    };
    
    void info() override
    {
        std::cout <<"Bus for passengers" << std::endl;
    }
};

class Truck: public Vehicle //класс наслідник №2
{
public:
    Truck(double fuel, double distance) : Vehicle( fuel, distance ) {} //заміна base()

    double fuel_consumption_calc() override   //override абстрактного методу
    {
       return m_fuel * m_distance / 100;
    };
    
    void info() override 
    {
        std::cout <<"Truck is working" << std::endl;
    }
       
};

int main()
{
    Bus bus(20, 40);
    std::cout << "fuel consumption: on distantion "<< bus.Getdistance() <<" km is: " << bus.fuel_consumption_calc() <<" liters" << std::endl;
    bus.info();
    std::cout << std::endl;

    Truck truck(120, 300);
    std::cout << "fuel consumption on distantion "<< truck.Getdistance() <<" km is: " << truck.fuel_consumption_calc() << " liters" << std::endl;
    truck.info();
    std::cout << std::endl;

    Vehicle* vehicles[] = {&bus, &truck};

    for(int i = 0; i<2; i++)
    {
         std::cout << "fuel consumption on distantion "<< vehicles[i]->Getdistance() <<" km is: " << vehicles[i]->fuel_consumption_calc() << " liters" << std::endl;
         vehicles[i]->info();
    }

}
